File: python/second/combination_sum_ii.py
class Solution(object):
    # O(n * 2^n) time | O(n) space
    def combinationSum2(self, candidates, target):
        """
        :type candidates: List[int]
        :type target: int
        :rtype: List[List[int]]
        """
        res = []
        part = []
        candidates.sort()

        def dfs(i, sum):
            if sum == target:
                res.append(part[::])
                return
            if sum > target or i >= len(candidates):
                return
            part.append(candidates[i])
            dfs(i + 1, sum + candidates[i])
            part.pop()
            while i + 1 < len(candidates) and candidates[i + 1] == candidates[i]:
                i += 1
            dfs(i + 1, sum)

        dfs(0, 0)
        return res


candidates = [10,1,2,7,6,1,5]
target = 8

solution = Solution()
print(solution.combinationSum2(candidates, target))

File: src/main/java/august_2025/CombinationSumII.java
package august_2025;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class CombinationSumII {

    public static void main(String[] args) {
//        int[] candidates = {2, 3, 6, 7};
//        int target = 7;

        int[] candidates = {10,1,2,7,6,1,5};
        int target = 8;

        CombinationSumII combinationSumII = new CombinationSumII();
        List<List<Integer>> result = combinationSumII.combinationSum2(candidates, target);
        System.out.println(result);
    }

    // O(n * 2^n) time | O(n) space
    public List<List<Integer>> combinationSum2(int[] candidates, int target) {
        Arrays.sort(candidates);
        List<List<Integer>> result = new ArrayList<>();
        List<Integer> part = new ArrayList<>();

        backtrack(candidates, 0, 0, target, part, result);

        return result;
    }

    private void backtrack(int[] candidates, int idx, int sum, int target, List<Integer> part, List<List<Integer>> result) {
        if (sum == target) {
            result.add(new ArrayList<>(part));
            return;
        }
        if (sum > target || idx >= candidates.length) {
            return;
        }
        part.add(candidates[idx]);
        backtrack(candidates, idx + 1, sum + candidates[idx], target, part, result);
        part.remove(part.size() - 1);
        while (idx + 1 < candidates.length && candidates[idx] == candidates[idx + 1]) {
            idx++;
        }
        backtrack(candidates, idx + 1, sum, target, part, result);
    }

}
